### ISSUE TITLE
[FEATURE] Vanilla Player Color Presets

### ISSUE THREAD

**Issue by OpenRift412** [odamex/odamex#322]
I think it would be kind of cool to see the ability to have player color presets in the player settings menu, akin to how the Eternity Engine does it. 

Here's an example of what I mean:
![image](https://user-images.githubusercontent.com/29780788/116296779-9e58f700-a768-11eb-94b7-38efd63d9bd9.png)

**Comment by Ch0wW** [comment]
That is actually a great idea, and would be a great complement on another "feature restoration" that is getting back the classic intermission screen! The idea to get a preset of the 4 playercolors and a 5th one that would be "custom" is actually something that shouldn't be too troublesome to make.

**Comment by Acts19quiz** [comment]
With 10.4.0, this can now be closed as done. Thanks everyone! (:

**Comment by OpenRift412** [comment]
> With 10.4.0, this can now be closed as done. Thanks everyone! (:

I was actually taking a look at this, and the player 2, 3, and 4 colors don't seem to match up properly with their vanilla counterparts. 

Here are some comparisons between Odamex and Zandronum, which has vanilla-accurate colors.

**Indigo**
Zandronum/Vanilla:
![image](https://github.com/odamex/odamex/assets/29780788/e7135528-ad8c-48fd-9c5b-b7d8a37b298f)
Odamex:
![image](https://github.com/odamex/odamex/assets/29780788/079a1913-f66a-4328-b527-f7f6455c6933)

**Brown**
Zandronum/Vanilla:
![image](https://github.com/odamex/odamex/assets/29780788/76a08eb9-eed0-46e6-a496-3543a6a91cfc)
Odamex:
![image](https://github.com/odamex/odamex/assets/29780788/1a405335-dca7-4022-a676-152ececa49d5)

**Red**
Zandronum/Vanilla:
![image](https://github.com/odamex/odamex/assets/29780788/37ab0205-62a9-47f4-a75c-5edd547d6dcc)
Odamex:
![image](https://github.com/odamex/odamex/assets/29780788/7b5cc083-ff6d-49b1-b744-bd19a2d1b2d1)

**Comment by Acts19quiz** [comment]
Ah, yes. That. There was a conversation about this in the Discord server when I first started this. Some copy/paste:

I have tweaked them some, though outright Vanilla matching might not be possible as-is.

Looking into Chocolate Doom, the way Vanilla does this is just take the colours of the green pixel range, and shift all of those palette values towards the desired place in the DOOM palette. These customize sliders like in ZDoom and Odamex just take green and mutate it to roughly match the chosen color. However, you're not getting direct palette value-to-palette value matching, but a translation of all green pixels. Thus you can get goofy results sometimes, like gray pixels standing out, or weird aberrations like bright (saturated) pink smears in the 'pink' palette range. Probably the biggest issue is the legs are invariably darker than they should be, especially compared to the arms.

Things definitely look worse in 8-bit than 32-bit. Nevertheless, there might not be a 'perfect' "ba b9 b9" type of value that lands on exactly the way Vanilla does it.

Modern GZDoom gets around this by having the preset values refer to a direct pixel-to-pixel mapping. That might be a good 'next stage' after the feature gets launched and completed in its first stage. However, that system within the draw system is an entirely separate setup, and Player Setup is already full of duct taping (from what I can tell) as is.

tl;dr Changing colours based on RGB values rather than vanilla palette index isn't going to result in perfect colour matching. It's always going to look a bit off. There's always room to tweak the default RGB values in the code (that's a simple fix), but it's still always going to look a tad off in certain areas. E.g., making the pants the correct brightness will make the arms way too bright. At this point, it's just trial and error on the hex values.

**Comment by Acts19quiz** [comment]
![PColor_Odamex_Zandronum](https://github.com/odamex/odamex/assets/63941237/dead4148-236e-4f97-ac47-2d7874683200)
There's a side-by-side I got going while trying to use the custom slider to match Zandronum/vanilla. The shoulders are too bright, but the chest and in particular the pants are too dark. This is because RGB values are being translated to match over the green pixels, whereas in vanilla, it's just shifting the colour palette to the proper indices. Fixing this would require an overhaul to how player colours are internally drawn. It's the same reason why no amount of tinkering with the sliders will produce a green exactly like the one the sprites themselves are.

But at the least now, the option framework is there to begin considering and planning bigger changes to how player customizations are drawn.

**Comment by OpenRift412** [comment]
> Ah, yes. That. There was a conversation about this in the Discord server when I first started this. Some copy/paste:
> 
> I have tweaked them some, though outright Vanilla matching might not be possible as-is.
> 
> Looking into Chocolate Doom, the way Vanilla does this is just take the colours of the green pixel range, and shift all of those palette values towards the desired place in the DOOM palette. These customize sliders like in ZDoom and Odamex just take green and mutate it to roughly match the chosen color. However, you're not getting direct palette value-to-palette value matching, but a translation of all green pixels. Thus you can get goofy results sometimes, like gray pixels standing out, or weird aberrations like bright (saturated) pink smears in the 'pink' palette range. Probably the biggest issue is the legs are invariably darker than they should be, especially compared to the arms.
> 
> Things definitely look worse in 8-bit than 32-bit. Nevertheless, there might not be a 'perfect' "ba b9 b9" type of value that lands on exactly the way Vanilla does it.
> 
> Modern GZDoom gets around this by having the preset values refer to a direct pixel-to-pixel mapping. That might be a good 'next stage' after the feature gets launched and completed in its first stage. However, that system within the draw system is an entirely separate setup, and Player Setup is already full of duct taping (from what I can tell) as is.
> 
> tl;dr Changing colours based on RGB values rather than vanilla palette index isn't going to result in perfect colour matching. It's always going to look a bit off. There's always room to tweak the default RGB values in the code (that's a simple fix), but it's still always going to look a tad off in certain areas. E.g., making the pants the correct brightness will make the arms way too bright. At this point, it's just trial and error on the hex values.

Well I think the palette translation method would be a nice thing to see at least for the OG colors at some point down the road. Maybe the same for the additional color presets in the same manner that the Eternity engine offers.

**Comment by bcahue** [comment]
Completed via #1594?

**Comment by rakohus** [comment]
Yes. Completed via 1594
